File: testdata/precision-suite-php/tp_ssrf.php
<?php
// SSRF: a user-supplied URL is fetched via cURL with no host allowlist, so an
// attacker can make the server request internal endpoints. CWE-918, TAINT-006.

function proxy()
{
    $url = $_GET['url'];
    $ch = curl_init($url);
    curl_setopt($ch, CURLOPT_RETURNTRANSFER, true);
    return curl_exec($ch); // nox-expect: TAINT-006
}

File: testdata/precision-suite-java/tp_ssrf.java
// SSRF: a user-supplied URL is fetched server-side with no host allowlist, so
// an attacker can reach internal services (169.254.169.254, localhost admin
// ports) — CWE-918. A correct scanner fires TAINT-006.
package com.example.proxy;

import java.io.InputStream;
import java.net.URL;
import javax.servlet.http.HttpServletRequest;

public class UrlProxy {

    // fetch opens a connection to an attacker-controlled URL — the vulnerability.
    public InputStream fetch(HttpServletRequest request) throws Exception {
        String target = request.getParameter("url");
        return new URL(target).openStream(); // nox-expect: TAINT-006
    }
}
